cli/tools/fmt.rs

```diff
@@ -223,23 +278,84 @@ pub fn format_json(
   dprint_plugin_json::format_text(file_path, file_text, &config)
 }
 
-/// Formats a single TS, TSX, JS, JSX, JSONC, JSON, or MD file.
+/// Formats IPYNB files.
+///
+/// attempt to format `code` cells if metadata indicates it's TypeScript.
+fn format_ipynb(
+  ipynb: Ipynb,
+  fmt_options: &FmtOptionsConfig,
+) -> Result<Option<String>, AnyError> {
+  fn source_from(text: String) -> Source {
+    Source::Array(text.split_inclusive('\n').map(|x| x.into()).collect())
+  }
+
+  let is_typescript = ipynb.is_typescript();
+  let cells: Vec<Result<Cell, Cell>> = ipynb
+    .cells
+    .into_iter()
+    .map(|cell| {
+      let source = match cell.source {
+        Source::String(text) => text,
+        Source::Array(texts) => texts.join(""),
+      };
+
+      let formatted = match cell.cell_type.as_str() {
+        "code" if is_typescript => Some("temp.ts"),
+        "markdown" => Some("temp.md"),
+        _ => None,
+      }
+      .map(|path| format_file(&PathBuf::from(path), &source, fmt_options))
+      .unwrap_or(Ok(None));
+
+      match formatted {
+        Ok(Some(text)) => Ok(Cell {
+          source: source_from(text),
+          ..cell
+        }),
+        _ => Err(Cell {
+          source: source_from(source),
+          ..cell
+        }),
+      }
+    })
+    .collect();
+
+  let already_formatted = cells.iter().all(|cell| cell.is_err());
+
+  if already_formatted {
```

Comment:
We could keep the JSON formatting exactly the same as the input by using `jsonc_parser::parse_to_ast` and then creating a list of [`deno_ast::TextChange`s](https://docs.rs/deno_ast/0.31.5/deno_ast/struct.TextChange.html) to apply, then apply them with [`deno_ast::apply_text_changes`](https://docs.rs/deno_ast/0.31.5/deno_ast/fn.apply_text_changes.html).

Reply:
@scarf005 so tonight I had a strong urge to add a dprint jupyter plugin that formats code blocks using other dprint plugins and I wrote it from scratch using jsonc_parser like how I described because it wasn't a lot of code to do. I feel slightly bad though because I realize now that it would be trivial to call the plugin's rust API and it makes too much sense to unify this code (https://docs.rs/dprint-plugin-jupyter/0.1.1/dprint_plugin_jupyter/fn.format_text.html -- format_with_host is to format a codeblock). Sorry about that, but at least this will speed up landing this. Would you be willing to update this code to import that crate and call that function instead? I didn't extensively test it so perhaps there are bugs.

Reply:
sure, no problem.

Reply:
Thanks and sorry about that once again. If you notice any issues or improvements that could be made, please open a PR there. The code is here: https://github.com/dprint/dprint-plugin-jupyter/blob/main/src/format_text.rs